### PR TITLE
[Challenge-1] Implemented the first REST endpoint that allows anyone …

### PR DESCRIPTION
**Pull request by luizamujdei**
…to publish a letter to Santa + tests

### CODE REVIEW

File: letter-sender/src/main/java/com/christmas/letter/sender/controller/ChristmasLetterController.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  private String topicArn;
+
+  public ChristmasLetterController(SnsClient snsClient){
+    this.snsClient = snsClient;
+  }
+
+  @PostMapping("/send")
+  public String sendChristmasLetter(@RequestBody ChristmasLetter letter) {
+
+    String message = String.format("Email: %s\nName: %s\nWishes: %s\nLocation: %s",
```
Comment (teodora-bobirneci) [inline letter-sender/src/main/java/com/christmas/letter/sender/controller/ChristmasLetterController.java]: I wouldn't want my controller to directly invoke the sns client. Imagine we had more complex things to do here, they don't belong in the controller. Also, you can send the letter directly as a json representation, no need to reformat etc.

File: letter-sender/src/main/java/com/christmas/letter/sender/controller/SampleController.java [hunk 2]
```diff
@@ -1,4 +1,4 @@
-package com.christmas.letter.sender;
+package com.christmas.letter.sender.controller;
```
Comment (teodora-bobirneci) [inline letter-sender/src/main/java/com/christmas/letter/sender/controller/SampleController.java]: We can delete this if we don't use it anymore.

File: letter-sender/src/test/java/SendChristmasLetterTest.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+                .endpointOverride(URI.create(LOCALSTACK_URL))
+                .credentialsProvider(StaticCredentialsProvider.create(AwsBasicCredentials.create(DUMMY, DUMMY)))
+                .region(Region.of(REGION))
+                .build();
+    }
+
+    @Test
+    public void whenSendChristmasLetter_ThenMessageReceivedOnQueue() throws IOException {
+        //Publish message to the SNS Topic
+        String testMessage = "{\n" +
```
Comment (teodora-bobirneci) [inline letter-sender/src/test/java/SendChristmasLetterTest.java]: You can use text blocks for increased readability. IntelliJ knows how to do it automatically. Just Alt + Enter => magic

File: letter-sender/src/test/java/SendChristmasLetterTest.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+                "    \"email\": \"luiza.mujdei@gmail.com\",\n" +
```
Comment (teodora-bobirneci) [inline letter-sender/src/test/java/SendChristmasLetterTest.java]: If this is your real email, i wouldn't post it in a public github repo.

File: letter-sender/src/test/java/SendChristmasLetterTest.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    private static final String LOCALSTACK_ARN = "arn:aws:sqs:us-east-1:000000000000";
+    private static final String REGION = "us-east-1";
+
+    private SnsClient snsClient;
+    private SqsClient sqsClient;
+    private String topicArn;
+    private String queueURL;
+
+    @BeforeAll
+    void setup(){
```
Comment (teodora-bobirneci) [inline letter-sender/src/test/java/SendChristmasLetterTest.java]: Most likely, if we would have more complex functionalities, we would need to kind of have the same setup for the other tests as well, so it's worth extracting this into a new class that can be reused.

File: letter-sender/src/test/java/SendChristmasLetterTest.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+                "    \"email\": \"luiza.mujdei@gmail.com\",\n" +
+                "    \"name\": \"Luiza\",\n" +
+                "    \"wishes\": \"fimo\",\n" +
+                "    \"location\": \"Iasi\"\n" +
+                "}";
+
+        snsClient.publish(PublishRequest.builder().topicArn(topicArn).message(testMessage).build());
+
+        //Polling the SQS queue
+        ReceiveMessageResponse receiveMessageResponse = sqsClient.receiveMessage(ReceiveMessageRequest.builder()
```
Comment (teodora-bobirneci) [inline letter-sender/src/test/java/SendChristmasLetterTest.java]: Maybe we can simplify this a little bit. If you would post the message as json, maybe you can deserialize it back in a ChristmasLetter instance and compare it to what you were expecting.

Also, what you're testing here is that snsClient knows how to publish a message, and sqsClient knows how to receive it. We should be testing our controller instead. See maybe MockMvc as a solution, example: https://spring.io/guides/gs/testing-web

File: letter-sender/src/test/java/SendChristmasLetterTest.java [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+public class SendChristmasLetterTest {
```
Comment (teodora-bobirneci) [inline letter-sender/src/test/java/SendChristmasLetterTest.java]: Let's also add tests for negative scenarios. Maybe the email is empty or invalid, etc

File: letter-sender/src/main/java/com/christmas/letter/sender/controller/ChristmasLetterController.java (hunk 1 above)
Comment (luizamujdei) [reply letter-sender/src/main/java/com/christmas/letter/sender/controller/ChristmasLetterController.java]: You are right, I added a service layer

File: letter-sender/src/test/java/SendChristmasLetterTest.java (hunk 6 above)
Comment (luizamujdei) [reply letter-sender/src/test/java/SendChristmasLetterTest.java]: I followed your suggestion and added a ChristmasLetterControllerTest that uses WebMvcTest

File: letter-sender/src/test/java/SendChristmasLetterTest.java (hunk 7 above)
Comment (luizamujdei) [reply letter-sender/src/test/java/SendChristmasLetterTest.java]: I added a negative test for an invalid request message

File: letter-sender/src/main/java/com/christmas/letter/sender/service/ChristmasLetterService.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+            //Publish to the SNS topic
+            PublishRequest request = PublishRequest.builder()
+                    .topicArn(topicArn)
+                    .message(message)
+                    .build();
+
+            PublishResponse response = snsClient.publish(request);
+
+            return "Message published to the  topic with message id" + response.messageId();
```
Comment (tudorcodrea) [inline letter-sender/src/main/java/com/christmas/letter/sender/service/ChristmasLetterService.java]: there is another option of pushing events by leveraging SnsTemplate and convertAndSend which has the payload conversion behind either way looks good

File: letter-sender/src/main/java/com/christmas/letter/sender/config/AwsConfig.java [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+
+@Configuration
+public class AwsConfig {
+
+    @Bean
+    public SnsClient snsClient(){
+        return SnsClient.builder()
+                .region(Region.US_EAST_1)
+                .endpointOverride(URI.create("http://localhost:4566"))
+                .build();
```
Comment (tudorcodrea) [inline letter-sender/src/main/java/com/christmas/letter/sender/config/AwsConfig.java]: all the properties for the external dependent resources is recommended to be loaded from properties files. I would leverage application.properties for this

**Review by tudorcodrea — COMMENTED** [review]
left a couple of comments. looks verry good

File: letter-sender/src/main/java/com/christmas/letter/sender/service/ChristmasLetterService.java (hunk 8 above)
Comment (luizamujdei) [reply letter-sender/src/main/java/com/christmas/letter/sender/service/ChristmasLetterService.java]: Thanks for the tip, I changed it to use SnsTemplate

File: letter-sender/src/main/java/com/christmas/letter/sender/config/AwsConfig.java (hunk 9 above)
Comment (luizamujdei) [reply letter-sender/src/main/java/com/christmas/letter/sender/config/AwsConfig.java]: you're right, fixed it

File: letter-sender/src/test/java/com/christmas/letter/sender/controller/ChristmasLetterControllerTest.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+        //Mock sending the message
+        mockMvc.perform(post(PATH)
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .content(testMessage))
+                .andExpect(status().isCreated());
+    }
+
+    @Test
+    public void whenSendChristmasLetter_withInvalidRequest_ThenReturnBadRequest() throws Exception {
```
Comment (teodora-bobirneci) [inline letter-sender/src/test/java/com/christmas/letter/sender/controller/ChristmasLetterControllerTest.java]: In a production application, I would expect to have a test for each parameter that is invalid. Also, let's also add checks for the validation details, example "email is required", "email is invalid" etc. If you're not familiar with the concept, and if you are, I would suggest you to look into parameterized tests. 
https://www.baeldung.com/parameterized-tests-junit-5

For me, CsvSource is the simplest method, but you can use whatever you see fit.

File: letter-sender/src/main/java/com/christmas/letter/sender/model/ChristmasLetter.java [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+    @NotEmpty(message = "Name is a mandatory field")
+    private String name;
+
+    @NotEmpty(message = "Wishes is a mandatory field")
+    private String wishes;
+
+    @NotEmpty(message = "Location is a mandatory field")
+    private String location;
+
+    public String getEmail() {
```
Comment (teodora-bobirneci) [inline letter-sender/src/main/java/com/christmas/letter/sender/model/ChristmasLetter.java]: You can use lombok to get rid of all the builerplate code.

**Review by teodora-bobirneci — APPROVED** [review]

**Review by tudorcodrea — APPROVED** [review]

File: letter-sender/src/main/java/com/christmas/letter/sender/model/ChristmasLetter.java (hunk 11 above)
Comment (luizamujdei) [reply letter-sender/src/main/java/com/christmas/letter/sender/model/ChristmasLetter.java]: added lombok and removed the unnecessary methods

File: letter-sender/src/test/java/com/christmas/letter/sender/controller/ChristmasLetterControllerTest.java (hunk 10 above)
Comment (luizamujdei) [reply letter-sender/src/test/java/com/christmas/letter/sender/controller/ChristmasLetterControllerTest.java]: I parameterized the test, also added extra validation message
